runners/flink/src/main/java/org/apache/beam/runners/flink/FlinkStreamingPortablePipelineTranslator.java

```diff
@@ -0,0 +1,443 @@
+                    windowingStrategy.getWindowFn().windowCoder());
+
+    TypeInformation<WindowedValue<KV<K, Iterable<V>>>> outputTypeInfo =
+            new CoderTypeInformation<>(outputCoder);
+
+    TupleTag<KV<K, Iterable<V>>> mainTag = new TupleTag<>("main output");
+
+    // TODO: remove non-portable operator re-use
+    WindowDoFnOperator<K, V, Iterable<V>> doFnOperator =
+            new WindowDoFnOperator<K, V, Iterable<V>>(
+                    reduceFn,
+                    pTransform.getUniqueName(),
+                    (Coder) windowedWorkItemCoder,
+                    mainTag,
+                    Collections.emptyList(),
+                    new DoFnOperator.MultiOutputOutputManagerFactory(mainTag, outputCoder),
+                    windowingStrategy,
+                    new HashMap<>(), /* side-input mapping */
+                    Collections.emptyList(), /* side inputs */
+                    context.getPipelineOptions(),
+                    inputElementCoder.getKeyCoder());
+
+    SingleOutputStreamOperator<WindowedValue<KV<K, Iterable<V>>>> outputDataStream =
+            keyedWorkItemStream
+                    .transform(
+                            pTransform.getUniqueName(),
+                            outputTypeInfo,
+                            (OneInputStreamOperator) doFnOperator);
+
+    context.addDataStream(
+            Iterables.getOnlyElement(pTransform.getOutputsMap().values()),
+            outputDataStream);
+
+  }
+
+  private void translateImpulse(
+          String id,
+          RunnerApi.Pipeline pipeline,
+          StreamingTranslationContext context) {
+    RunnerApi.PTransform pTransform =
+            pipeline.getComponents().getTransformsOrThrow(id);
+
+    TypeInformation<WindowedValue<byte[]>> typeInfo =
+            new CoderTypeInformation<>(
+                    WindowedValue.getFullCoder(ByteArrayCoder.of(), GlobalWindow.Coder.INSTANCE));
+
+    DataStreamSource<WindowedValue<byte[]>> source = context.getExecutionEnvironment()
+            .fromCollection(Collections.singleton(WindowedValue.valueInGlobalWindow(
+                    new byte[0])), typeInfo);
+
+    context.addDataStream(
+            Iterables.getOnlyElement(pTransform.getOutputsMap().values()),
+            source);
+  }
+
+  private <T> void translateAssignWindows(String id, RunnerApi.Pipeline pipeline,
+                                          StreamingTranslationContext context) {
+    RunnerApi.Components components = pipeline.getComponents();
+    RunnerApi.PTransform transform = components.getTransformsOrThrow(id);
+    RunnerApi.WindowIntoPayload payload;
+    try {
+      payload = RunnerApi.WindowIntoPayload.parseFrom(transform.getSpec().getPayload());
+    } catch (InvalidProtocolBufferException e) {
+      throw new IllegalArgumentException(e);
+    }
+    WindowFn<T, ? extends BoundedWindow> windowFn = (WindowFn<T, ? extends BoundedWindow>)
```

Comment:
This only works for well known window fns, we should defer this execution to the SDK if the WindowFn can't be parsed or just defer it all the time.

Python already supports WindowInto (https://github.com/apache/beam/blob/5b791489a5884cfd298578324ec08238d28f303c/sdks/python/apache_beam/runners/worker/bundle_processor.py#L556) while Java support is being added in https://github.com/apache/beam/pull/5329